File: 2022/august/10-08-2022/convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def sortedArrayToBST(self, nums):
        if len(nums) == 0:
            return None
        left, right = 0, len(nums)
        mid = (right-left) // 2
        root = TreeNode(nums[mid])
        root.left = self.sortedArrayToBST(nums[left:mid])
        root.right = self.sortedArrayToBST(nums[mid+1:right])
        return root


print(Solution().sortedArrayToBST([-10,-3,0,5,9]))

File: 2022/august/10-08-2022/convert-sorted-array-to-binary-search-tree.rs
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn sorted_array_to_bst(mut nums: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        let right = nums.split_off((nums.len() + 1) / 2);
        match nums.pop() {
            None => None,
            Some(val) => Some(Rc::new(RefCell::new(TreeNode {
                val,
                left: Self::sorted_array_to_bst(nums),
                right: Self::sorted_array_to_bst(right),
            })))
        }
    }
}
